CONTRIBUTING.md

```diff
@@ -1,3 +1,98 @@
+
+All Trainee methods and attributes are defined in either `trainee_template.amlg` or in one of the Amalgam
+files defined within the `trainee_template` directory. Attributes refer to the variables that hold cached values
+and other information that are relevant for reuse to the Trainee. Methods refer to the callable sections of code
+that execute machine learning or auxiliary tasks.
+
+# Development
+
+## Debugging
+
+Debugging the Howso Engine is supported through the debugging tools provided for the
+[Amalgam language](https://github.com/howsoai/amalgam). Specifically, the [Amalgam extension
+for VSCode](https://github.com/howsoai/amalgam-ide-support-vscode) is recommended for debugging the
+Howso Engine. This extension supports debugging Amalgam scripts as well as trace files produced
+by [amalgam-lang-py](https://github.com/howsoai/amalgam-lang-py). Please see these projects to
+get further information about debugging Amalgam.
+
+## Testing
+
+The Howso Engine has a suite of unit tests and performance tests. To run the full set of unit tests, simply
+call the desired Amalgam executable on `ut_comprehensive_unit_test.amlg` from within the `unit_tests` directory.
+
+Unit tests are held in the `unit_tests` directory while performance tests are held in the `performance_tests`
+directory.
+
+To create your own tests, we recommend creating a new Amalgam script in the `unit_tests` directory. It is
+also recommended to use the following code at the beginning of the test:
+
+```
+#unit_test (direct_assign_to_entities (assoc unit_test (load "unit_test.amlg")))
+(call (load "unit_test_howso.amlg") (assoc name "ut_your_custom_test_name.amlg"))
+```
+
+This code imports the unit testing framework defined within `unit_tests/unit_test_amlg` and calls the code
+within `unit_tests/unit_test_howso.amlg` which instantiates the `howso` entity, which can be used to call
+the API endpoints defined in `howso.amlg`. This unit testing framework includes methods such as `#assert_same`,
+`#assert_approximate`, and more. Please see `unit_tests/unit_test.amlg` to see the full list of methods to
+help with unit testing. It is also recommended to add periodic calls to `#exit_if_failures` in longer test
+files so that tests can fail out early. The `#assert_*` methods will **not** halt the execution of a test alone.
+
+If you wish to add a test to the comprehensive test set, you can add it to the appropriate test file that is
+called from `ut_comprehensive_unit_test.amlg`. If you want to test an endpoint, it makes the most sense to
+add your test to `ut_howso.amlg`. If you want to test a Trainee method, then it makes more sense to add
+your test to `ut_trainee_template.amlg`. Please read these files to see how to add your test file to the set.
+
+## Style Guide
+
+The styling of code written in the Howso Engine inherits the styling of the Amalgam language with additions
+that help reading the code of the Engine a bit easier. These additions are mostly the use of different
```

Comment:
Grammar -- "...with additions that help reading the code of the Engine a bit easier"